safehttp/plugins/xsrf/xsrf.go

```diff
@@ -63,24 +56,52 @@ func Token(r *safehttp.IncomingRequest) (string, error) {
 	return tok.(string), nil
 }
 
+func addTokenCookie(w *safehttp.ResponseWriter) (*safehttp.Cookie, error) {
+	buf := make([]byte, 20)
+	_, err := rand.Read(buf)
+	if err != nil {
+		panic(fmt.Errorf("failed to generate entropy using crypto/rand/RandReader: %v", err))
+	}
+
+	c := safehttp.NewCookie(tokenCookie, base64.StdEncoding.EncodeToString(buf))
+	c.SetSameSite(safehttp.SameSiteStrictMode)
+	if err := w.SetCookie(c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
 // Before should be executed before directing the safehttp.IncomingRequest to
 // the handler to ensure it is not part of the Cross-Site Request
 // Forgery attack.
 //
+// On first user visit through a state preserving request (GET, HEAD or
+// OPTIONS), it sets a nonce-based cookie used to identify a user. This will be
+// used in the token generation and verification algorithm and is expected in
+// all subsequent incoming requests.
+//
 // In case of state changing requests (all except GET, HEAD and OPTIONS), it
 // checks for the presence of an XSRF token in the request and validates it
-// based on the user ID associated with the request.
+// based on the custom cookie.
 //
 // For authorized requests, it adds a cryptographically safe XSRF token to the
-// incoming request. It can be later extracted using Token.
+// incoming request. This can be later extracted using Token.
 func (i *Interceptor) Before(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, cfg interface{}) safehttp.Result {
-	userID, err := i.Identifier.UserID(r)
```

Comment:
I don't get this part. Why is this removed?

Reply:
So, the previous plugin version was vulnerable to pre-login CSRF. An attack would go as follows: I am logged into your bank account and the attacker also has an actual account on the banking website. Then, the attacker can forge a login request in my browser and trick me into using their account. Then, they'll have access to stuff like my transaction history while I am logged in into their account.
 
This would be allowed by the previous plugin because the call to `Identifier.UserID(r)` request would return the ID corresponding to the attacker account even if I am the one navigating and the token that is injected in the form has also been generated for the attacker's account.

Therefore, we need a way to identify that I am the one navigating the attacker's account. One way to do this is: the first time I navigate the banking website with a state preserving request, a noonce-based cookie is generated with SameSite Strict. Then, any xsrf token validation fails because the token was generated with the attacker's cookie and we validate with our cookie.

Hopefully this made sense (@empijei please correct me if I misunderstood something).

Reply:
But now you've taken away the feature of extracting a user ID from a request, possibly via some lookup operations.

Reply:
AFAIU that is not needed anymore.

Reply:
Discussed this with @empijei offline.

Long story short:
- you're right, we don't really need the `UserIdentifier` - it wasn't clear to me why
- please document why need both the cookies + token injection in the response